superset/utils/excel.py

```diff
@@ -19,16 +19,36 @@
 
 import pandas as pd
 
+from superset.utils.core import GenericDataType
 
-def df_to_excel(df: pd.DataFrame, **kwargs: Any) -> Any:
+
+def df_to_excel(
+    df: pd.DataFrame, coltypes: list[GenericDataType], **kwargs: Any
+) -> Any:
     output = io.BytesIO()
 
     # timezones are not supported
     for column in df.select_dtypes(include=["datetimetz"]).columns:
         df[column] = df[column].astype(str)
 
+    ndf = convert_df_with_datatypes(df, coltypes)
+
     # pylint: disable=abstract-class-instantiated
     with pd.ExcelWriter(output, engine="xlsxwriter") as writer:
-        df.to_excel(writer, **kwargs)
+        ndf.to_excel(writer, **kwargs)
 
     return output.getvalue()
+
+
+def convert_df_with_datatypes(
```

Comment:
I believe we could make it more explicit that the `DataFrame` can be modified and avoid the copy overhead. To do that, we need to separate column conversion from exporting. Something like:

```
def df_to_excel(df: pd.DataFrame, **kwargs: Any) -> bytes:
    output = io.BytesIO()

    # pylint: disable=abstract-class-instantiated
    with pd.ExcelWriter(output, engine="xlsxwriter") as writer:
        df.to_excel(writer, **kwargs)

    return output.getvalue()


def apply_column_types(
    df: pd.DataFrame, column_types: list[GenericDataType]
) -> pd.DataFrame:
    for column, column_type in zip(df.columns, column_types):
        if column_type == GenericDataType.NUMERIC:
            df[column] = pd.to_numeric(df[column])
        elif pd.api.types.is_datetime64tz_dtype(df[column]):
            # timezones are not supported
            df[column] = df[column].astype(str)
        else:
            df[column] = df[column].astype(column_type)
    return df
```

Reply:
I'm probably over-conservative but  I have two concerns

- `pd.to_numeric` may raise exception, from documentation the "ignore error" param is deprecated and it's recommended to catch exceptions. 

If everything is coherent from end to end, of course conversion will succeed. If you're confident it's the case and try/catch is useless that's fine for me. Else changing df in place is an issue.

- `else : df[column] = df[column].astype(column_type)` ... again I'm a bit scared of the effect on excel export side, because I don't know how it works. Id' rather "leave untouched" data not to be converted, but here again you probably know better.


Finally if I get things right, the call to "apply_column_types" should be made in query_context_processor, right ?

Reply:
Addendum : a quick test regarding the "else" conversion, it fails with this error

```
TypeError: Cannot interpret '<GenericDataType.STRING: 1>' as a data type
```

Reply:
> pd.to_numeric may raise exception, from documentation the "ignore error" param is deprecated and it's recommended to catch exceptions.

Good point. Here we have two options:
- Catch and convert to a default. We also need to alert the user that some values were modified when exporting.
- Don't catch and show a good message clearly indicating the column/row that is preventing the export.

> else : df[column] = df[column].astype(column_type) ... again I'm a bit scared of the effect on excel export side, because I don't know how it works. Id' rather "leave untouched" data not to be converted, but here again you probably know better.

I'm not sure what you mean by untouched data but I'm assuming that when exporting users want the data to comply with what is defined as the column type. That was your original problem for numeric types, so I'm assuming that other non-numerical columns that are not matching their defined types should also be converted.

Let me ask for @betodealmeida and @eschutho input here as they have more context on this feature and might help us.

Reply:
As en end user, I much more prefer having an xlsx file with data I have to manually reformat, than having nothing at all.

> Catch and convert to a default. We also need to alert the user that some values were modified when exporting.

probably same as "dates with timezone" fix : `df[column].astype(str)`

Alerting the user is faaaaaaaar beyond anything I would be able to do, and agani as an end user I think I would'nt care. Nothing is said about timezone conversion either and I bet nobody cares.

>Don't catch and show a good message clearly indicating the column/row that is preventing the export.

As said, as an end user I would rather have a not-perfectly formatted file, but will all data in it, than no data at all.


- About "untouched fields" :

I don't really remember all the details (because it dates a bit and it's not my passion tbh :) ), but at the time, I manually opened the produced xlsx files, meaning : unzip it and see the xml-ish data inside and try to understand the differences when converted data or not. From there, open in libreoffice or MS and see the behaviors, and try to find the best.
Numeric values are stored "in the cells", while others have references in the cell and values mapped somewhere else. Or maybe it's the other way'round - can't even remember which is which. Anyway there's a clear separation  between numeric and 'other'. Then, inside the 'other' category of course there may be further differences. 

From my understanding then, the best thing one can do to expect any office software to deal with these files is to make sure numerics are numerics, anything doesn't matter that much and autodetection of strings work.

That's probably why `df[column] = df[column].astype(str)` in the date conversion has no real bad effect on usability.

That's all I know, which is not much for such a long text :)